tlatools/org.lamport.tlatools/src/tla2sany/semantic/ModuleNode.java

```diff
@@ -419,6 +426,20 @@ public final OpDefNode[] getOpDefs() {
     }
     return opDefs;
   }
+  
+  /**
+   * Function {@link ModuleNode#getOpDefs} creates a cache of all operator
+   * definitions on first call. While this was fine when we could expect SANY
+   * to never have its parse tree modified after original parsing, now we
+   * occasionally want to insert a definition or expression & evaluate it in
+   * the context of an existing spec without having to restart the entire
+   * parse process. Once we do that we need to invalidate the cache so the
+   * definition can be surfaced when the cache is rebuilt; that is what this
+   * function does.
+   */
+  public void clearOpDefCache() {
```

Comment:
Are you certain that the underlying `ModuleNode#ctxt` cannot be mutate after `getOpDefs` is invoked the first time by legacy code?

Reply:
I'm not sure; this is something to explore with this prototype. The role of the context class is still a bit mysterious to me and I don't know how it connects to other things in the parser. It is home to a piece of static global state I want to get rid of though, so understanding it is hopefully in my future.